Fix img.scale for scaled integer data (#4600, #4588) - alternative to #4602

bamford opened the pull request:
Avoid exceptions in img.scale for scaled integer data and float BZERO,
and also for compressed scaled integer (e.g. uint) data.

The fix casts bzero to an integer type whenever that does not involve
a loss of precision (i.e. when bzero is a float type representation of
an integer).  An exception will still arise when using a non-integer
bzero while insisting on an integer data array, as the correct
behaviour is uncertain.  See #4602 for discussion.  The cast to
integer occurs after the original bzero is stored, so that checksum
tests still pass.

New regression tests are included.  All existing tests pass.

mhvk commented:
Looks good!

bamford commented:
I added a test to ensure the exception still got raised if non-integer bzero was specified with integer data, but it turns out the exceptions is only raised in numpy >= 1.10.  (+1 for tests!)  For earlier versions of numpy the in-place addition was dropping precision, and just issuing a DepreciationWarning!  So, actually PR #4602 was preserving the earlier behaviour.

I'll amend this part of the test so it passes on older numpy.

Then it's up to a higher power to decide whether to preserve the old behaviour (#4602) or throw an TypeException in the case on non-integer BZERO with integer data.

astrofrog commented:
@embray - if you have time, could you let us know if this seems reasonable?

embray commented:
Looks good to me. Should have changelog entry(ies) for next bugfix release(s)

bamford commented:
Thanks for looking at this.  I've put writing the changelog entry on my todo list.

anntzer commented:
Any chance either this or #4602, together with the additional fix I mentioned there (https://github.com/astropy/astropy/pull/4602#issuecomment-208713748), can make it to 1.2?

astrofrog commented:
@anntzer - since neither this nor #4602 were quite finished, we won't be able to include them in 1.2 - however we can certainly include them in 1.2.1 shortly after.

bamford commented:
I've added the [test suggested](https://github.com/astropy/astropy/pull/4602#issuecomment-208713748) by @anntzer and addressed this issue in a simple way (again, as suggested by @anntzer), which doesn't break any existing tests.

I've also included a brief changelog entry:
> Avoid new exceptions with numpy >= 1.10 when using scaled integer data: where BZERO has float type but integer value or where data is unsigned int.

Rebased on master in preparation for merging.

bamford commented:
On the dev email list I see there are may be plans for serious modifications to the way bzero and bscale are handled.  I've no problem if these PRs get thrown out as a result, but if that work is delayed, maybe they will help in the interim!

pllim commented:
@bamford , if possible, could you please link the relevant mailing list discussions mentioned above here, just for the record?

mhvk commented on astropy/io/fits/tests/test_image.py:
```diff
@@ -19,6 +19,7 @@
 
 from . import FitsTestCase
 
+NUMPY_LT_1P10 = [int(x) for x in np.__version__.split('.')[:2]] < [1, 10]
```
We try to collect these in one place:
```
from ....utils.compat import NUMPY_LT_1_10
```

mhvk commented on astropy/io/fits/tests/test_image.py:
```diff
@@ -992,6 +993,50 @@ def test_scaled_image_fromfile(self):
+        hdu0 = fits.PrimaryHDU(data=a.copy())
+        # Ensure this behaviour is preserved
+        if not NUMPY_LT_1P10:
+            pytest.raises(TypeError, hdu0.scale, 'int16', bzero=99.9)
```
I find this test very difficult to read... Could you rewrite it as follows?
```
with pytest.raises(TypeError):
    hd0.scale('int16', bzero=99.9)
```

mhvk commented on astropy/io/fits/tests/test_image.py:
```diff
@@ -1554,11 +1599,44 @@ def test_duplicate_compression_header_keywords(self):
+        hdu0 = fits.CompImageHDU(data=a.copy())
+        # Ensure this behaviour is preserved
+        if not NUMPY_LT_1P10:
+            pytest.raises(TypeError, hdu0.scale, 'int16', bzero=99.9)
```
Same here.

mhvk commented:
@bamford - A very few small comments, but otherwise this looks great!

MSeifert04 commented:
Didn't  #5053 adress this problem or parts of it?

bamford commented:
Ugh. I _thought_ I'd avoided breaking existing tests, but actually the it breaks as unsigned integer data test. I can't quickly find a fix and I'm going to have to leave this for now.  It feels like it is part of a bigger issue anyway, which others seem to have a better idea about.

p.s. thanks for the tips @mhvk

mwcraig commented:
Any PR that adds more tests for this section of the code is great, thanks for working on this @bamford!

mhvk commented:
The problem in the failure is that the bzero (2**63) cannot be represented as a signed `np.int64` and therefore the numpy `subtract` routine returns `float64` output.

It *might* be solved if one reverts to writing `data += -_zero` instead of `data -= _zero` (since `-2**63` *can* be represented as `np.int64`). If so, do add a note...

anntzer commented:
Seems that the following patch lets all the tests pass:
```
         # Do the scaling
         if _zero != 0:
             # 0.9.6.3 to avoid out of range error for BZERO = +32768
-            self.data -= _zero
+            if not np.can_cast(
+                    np.min_scalar_type(_zero), self.data.dtype, "same_kind"):
+                raise TypeError
+            self.data += self.data.dtype.type(-_zero)
             self._header['BZERO'] = _zero
         else:
             try:
```
(we need the explicit `can_cast` check to prevent floats from passing though.)

pllim commented:
In addition, this also needs a rebase to resolve conflicts.

MSeifert04 commented:
Will the content of the PR conflict with #5250? I think these touch similar issues?

embray commented:
Yes, #5250 addressed the casting issue already, so this should be rebased on top of it.

anntzer commented:
Kindly requesting that this PR or #4602 (similar issue) be merged by 1.3.  Thanks!

mwcraig commented:
@bamford -- any objections to my taking this PR over so we can get it in this week? I'd cherry-pick your contributions onto the latest master so that you get proper credit.

If I don't hear by Tuesday evening I'll go ahead in the interests of finalizing it.

mwcraig commented:
@anntzer -- thanks for the bump, this had fallen off my radar.

mwcraig commented:
@pllim -- should this be re-milestoned to 1.3?

mwcraig commented:
@MSeifert04 -- #5250 fixed some, but not all, of the problem, it seems.

@bamford -- in checking to see whether this had already been fixed, I've ended up pretty much doing the rebase. I can walk you through doing a force push of those changes to your branch or I can open a new PR (unless someone reading this has a better approach).

mwcraig commented:
Closing this and #4602. Please see #5527, which contains the fixes in this PR with some editing to account for the changes made in #5250.

@bamford, thanks again for your work on this!

bamford commented:
@mwcraig, sorry for not responding sooner and many thanks for finishing the job!
